### ISSUE TITLE
Dapr state store component is not handling FIRST_WRITE concurrency option correctly.

### ISSUE THREAD

**Issue by prtk-g** [dapr/components-contrib#994]
## Expected Behavior

If there are two concurrent requests to save state for the same key, with concurrency option as `FIRST_WRITE`, and etag omitted or null, only one of the requests should succeed and the other should fail with an exception. 

## Actual Behavior

One of the requests overrides the other and there is a lost update.

## Steps to Reproduce the Problem

1. Start a Redis database server
```bash
docker run -d -p 8001:6379 redis
```
2. Create a dapr state component file
```yml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: redis
spec:
  type: state.redis
  version: v1
  metadata:
  - name: redisHost
    value: localhost:8001
  - name: redisPassword
    value: ""
```

3.  Run Dapr with state component file in the current directory
```bash
dapr run  --components-path . --dapr-http-port 3500
```
4. Send 1st request to save value for a key, with concurrency option as `FIRST_WRITE`
```bash
curl --location --request POST 'http://localhost:3500/v1.0/state/redis' \
--header 'Content-Type: application/json' \
--data-raw '[
    {
        "key": "1",
        "value": "value-1",
        "options": {
            "concurrency": "first-write",
            "consistency": "strong"
        }
    }
]'
```
5. Send 2nd request to save a different value for the same key, again with concurrency option as `FIRST_WRITE`. 
```bash
curl --location --request POST 'http://localhost:3500/v1.0/state/redis' \
--header 'Content-Type: application/json' \
--data-raw '[
    {
        "key": "1",
        "value": "value-2",
        "options": {
            "concurrency": "first-write",
            "consistency": "strong"
        }
    }
]'
```
6. Send request to get the state for saved key 
```bash
curl --location --request GET 'http://localhost:3500/v1.0/state/redis/1' \
--data-raw ''
```
Result: 
```bash
"value-2"
```
## Note
As per discussion on [this](https://discord.com/channels/778680217417809931/778749797242765342/860418694773735444) post on discord, if the etag value is null or is omitted, dapr handles the request in last write wins pattern. But since this may result in some complexity for saving state for a non-existing key, `FIRST_WRITE` option should be given priority over etag's null value, for handling the concurrency for the request.

## Release Note

RELEASE NOTE: **FIX** 

RELEASE NOTE:

**Comment by artursouza** [comment]
Please, add conformance test to validate/fix the same in other state stores.

**Comment by daixiang0** [comment]
Split into sub-tasks and fix Redis first as #1010.
